### DISCUSSION TITLE
Make assignment operator overloadable for "transitional" purposes or add something similar to conversion operators that takes "assignee" instance in account

### DISCUSSION THREAD

**Discussion by Servail** [dotnet/csharplang#2773] [Language Ideas] [answered]
While working with C#, I've found that conversion operator is not enough for my needs.
I've got:

	public class FloatSource
	{
		public float value = 1f;
	}

	public class HubFloat
	{
		public float internalValue;
		public FloatSource primarySource;
		public FloatSource secondarySource;

		public float value
		{
			get => primarySource? primarySource.Value : (secondarySource? secondarySource.Value : internalValue);
			set => primarySource? primarySource.Value : (secondarySource? secondarySource.Value : internalValue);
		}

                public static implicit operator HubFloat(float value)
	}

    HubFloat hubFloat = new HubFloat();
    float someValue = 10f;

Then I set "hubFloat.primarySource" and "hubFloat.secondarySource" from somewhere.
Next **I want to handle "hubFloat" like a simple "float"**, i.e. I want

    hubFloat = someValue;

to be exactly the same as

    hub.value = someValue;

And since assignment operator overloading is forbidden, the only potential way I found was implicit conversion operator. But conversion operator didn't allow to do so, instead it creates new object (which is in fact nullifies FloatSources - unacceptable), and there's no handle to old object, because conversion is not what I need, obviously (I know). I certainly need _something similar_ to assignment overloading.

So I understand why

    public static overload operator =(HubFloat lhs, HubFloat rhs) {}
    public static overload operator =(HubFloat lhs, HubFloatDerived rhs) {}

are forbidden. I agree with it. But why we can't use it for "assignment" (translation, transition, you name it) between different (and non-derived) types? 

Technically imaginary

    public static overload operator =(HubFloat hub, float value)
    {
        var newHub = new FloatHub();
        newHub.value = value;
        hub = newHub;
    }

is the same as

    public static implicit operator HubFloat(float value)
    {
        var newHub = new HubFloat();
        newHub.value = value;
        return newHub;
    }

but with more potential (that is certainly needed, I googled!). On practice it's the same "hubFloat = someValue" anyway, then why not allow the use of "assignee" side? It didn't look like immutability reasons, more like lack of functional.

So in ideal I want to use it like that:

    public static overload operator =(HubFloat hub, float value) //note these are different types!
    {
        hub.value = value;
    }

Or maybe to have some "conversion extension", which looks as silly as it sounds, especially next to "assignment overloading":

    public static implicit operator HubFloat(this HubFloat hub, float value)
    {
        hub.value = value;
        return hub;
    }

Or maybe theres another approach I don't know of.

P.S.: Yes, I've red Eric Lippert's articles (and there's still a problem), I know about C++ (which sounds absurd to use just to avoid this particular flaw, while everything else are good in C#) and I definitely don't want to use member access (obviously).

**Comment by YairHalberstadt** [comment]
I don't think you've explained why `hub.value = someValue;` is not acceptable.

**Comment by Servail** [comment]
@YairHalberstadt
> 
> 
> I don't think you've explained why `hub.value = someValue;` is not acceptable.

There it is:
>I want to handle "hubFloat" like a simple "float"

They must be interchangeable by design.
>I definitely don't want to use member access (obviously)

Or there will be no problem.

**Comment by YairHalberstadt** [comment]
> I want to handle "hubFloat" like a simple "float"
I definitely don't want to use member access (obviously)

hubFloat is a mutable reference type, whereas as float is an immutable value type. However hard you try to cover this up, it will never work properly, and you will just cause yourself more trouble.

Instead of fighting the language, you'd be better off doing this in the idiomatic way for C#. Use  `hub.value = someValue;` and everything will work so much more easily.

Ok, now that we've got that out the way, why can't you overload the assignment operator in C#? Well I don't know, but I can make a pretty good guess.

Assignment is probably the single most commonly used operation in C#. It appears in practically every second line of code. It means one of two things in every single location:

1: put the thing on the right in the location referred to by the name on the left:

2: call a property setter with the value on the right as the parameter.

That covers 99.9% of what you could possibly want the assignment operator to mean.

This makes it really easy to reason about what an assignment means when reading code. If you're assigning to a local it can only mean one thing, if you're assigning to a member, which is probably slightly rarer, it can only mean one of two things, and it's trivial to find out which - just go to the definition of the member, and see if it's a property or a field.

If you add a third meaning (which would rarely be used), you now have two options to what assignment to a local means, and three options for assignment to a member. Neither is it as trivial to find out which - you can't just go to the definition of the thing being assigned, you've got to go to the definition for the type you're assigning it to, and check if it overloads assignment.

That's a lot of extra overhead to anyone who's reviewing/reading code for only marginal benefit in some very rare (and IMO dubious) cases.

Besides for this there are technical issues to solve around the design:

1. Does the overloaded assignment affect subtypes, or only the base type.

2. How can I assign the type normally? If I can't then how do I even define the overloaded assignment operator which will usually want to call assignment at some point?

**Comment by Servail** [comment]
@YairHalberstadt
I don't want to overload assignment operator, I want something similar to conversion operator (just with access to original instance) with all that insists:
1. No, only specified Types.
2. Overloading for these cases is still forbidden, of course. In other cases you can't assign two different non-derived types anyway.

All of this stated in the original text, though.

**Comment by YairHalberstadt** [comment]
> I don't want to overload assignment operator, I want something similar to conversion operator (just with access to original instance)

These two are semantically equivalent.

**Comment by Servail** [comment]
@YairHalberstadt
> These two are semantically equivalent.

That's it. I mentioned this. And this is the argument that something like this can exist.

**Comment by YairHalberstadt** [comment]
@Servail

Most times I've learnt a new language / framework, at the beginning I'm frustrated by the things it can't do, and I try my hardest to find ways around its limitations. This is sometimes successful, sometimes not, but always frustrating.

Eventually though, I've always learnt to understand the mentality of the language/framework, and it's way of doing things, and invariably I realize that I should instead do things in the idiomatic manner. Immediately my frustrations go away, and more than that, when I see what I used to try to do, I understand why it wouldn't have been sensible to do things that way.

I think if you try just using member access, as is standard and recommended in C#, things will work out fine.

**Comment by YairHalberstadt** [comment]
>> These two are semantically equivalent.

> That's it. I mentioned this. And this is the argument that something like this can exis

?

I was arguing that overloadable assignment is a bad idea. You responded that you don't want overloadable assignment, you want conversion operator with access to original instance. I said that's just a different name for overloadable assignment (they are semantically equivalent), so it remains a bad idea. I'm not sure I understand your response to that.

**Comment by Servail** [comment]
@YairHalberstadt
>I think if you try just using member access, as is standard and recommended in C#, things will work out fine.

No doubt. But one day interfaces were unable to contain default implementations too. [Soon they will](https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/).  
>that's just a different name for overloadable assignment

Not really.
>they are semantically equivalent

So as with conversion operators. And nobody argues. This is out of the issue scope anyway.

**Comment by HaloFour** [comment]
What you are asking to allow sounds like it was the main reason the assignment operator was left off of the list of overloadable operators.  If you want to mutate a reference you can do so through member access or a mutator method.

**Comment by YairHalberstadt** [comment]
You stated that your aim here is:

> I want to handle "hubFloat" like a simple "float"
I definitely don't want to use member access (obviously)

That simple won't work. Even if this feature was implemented, they'll still be plenty of cases where this will work incorrectly. 

For example you want be able to cast a List<hubValue> to a List<float> or vice versa. Similarly for all generics.

Overloading methods accepting a hubValue won't work correctly with floats.

Reflection won't work with this.

And countless more examples.

So it's much easier accepting that these two are fundamentally different, than fighting a losing battle trying to paint over the cracks.

**Comment by Servail** [comment]
@HaloFour then why there is conversion operators? And I actually appretiate that overloading is forbidden in mentioned cases. But it's forbidden in all cases for no reason. 

@YairHalberstadt
>plenty of cases where this will work incorrectly

I've mentioned these cases. In other cases, the thing I want is harmless as conversion operator (which in fairness is pretty harmful, because I can return some "ill" static object instead of newly constructed).

**Comment by YairHalberstadt** [comment]
Sorry I mistyped. I meant won't be able, not want be able

**Comment by HaloFour** [comment]
@Servail 

> then why there is conversion operators?

Because conversion between types is useful and it doesn't redefine the assignment.

**Comment by Servail** [comment]
@HaloFour
> Because conversion between types is useful and it doesn't redefine the assignment.

So do feature that I'm requesting for. It doesn't redefine, it just looks semantically equivalent.

**Comment by HaloFour** [comment]
@Servail 

> It doesn't redefine, it just looks semantically equivalent.

Looks and behaves, because it is redefining the assignment operator.

If you want to argue that the language should lift the restriction on overloading the assignment operator, fine.  But don't frame your proposal that it's somehow not overloading the assignment operator and thus any argument against overloading the assignment operator doesn't apply.

**Comment by Servail** [comment]
@HaloFour
Yes, in the first place, I don't think that overloading of assignment operator is something bad. But the goal is different. Again, I don't want to redefine assignment operator itself, because it breaks all the reference assignment base. I just want to propose "transitional" operator, similar to conversional. It doesn't touch reference assignment base (these cases still forbidden), it's only for completely unrelated types, which you can't assign together anyway (like bool b = float f). So it maybe looks, but actually doesn't behaves.

**Comment by HaloFour** [comment]
@Servail 

That's still overloading the reassignment operator, even if you restrict it to between different types.  And in my opinion the self-mutation behavior you've described falls under the umbrella of abuse for that operator.  I wouldn't consider it good practice for languages that allow such overloading.  Member access is the idiomatic syntax for the problem you're trying to solve.

**Comment by Servail** [comment]
@HaloFour 
Devil is not so black as he is painted - otherwise C++ would be a bad lang. Therefore if something should happen, it will. As I said I can pass static object in conversion operator, so technically I can pass even same object and make it in fact mutate itself. Looks and behaves. But that's just words, and problem still exists. And it's not an idiomatic cost, it appears like a snag while everything else is good (in C#).

**Comment by Servail** [comment, marked as answer]
Black magic for all the non-believers:

    public class HubFloat
    {
        public HubFloat current = new HubFloat();

        public float internalValue;
        public FloatSource primarySource;
        public FloatSource secondarySource;

        public float value
        {
            get 
            {
                current = this;
                return primarySource!=null? primarySource.Value : (secondarySource!=null? secondarySource.Value : internalValue);
            }
            set
            {
                current = this;
                if (primarySource!=null) primarySource.Value = value;
                else if (secondarySource!=null) secondarySource.Value = value;
                else internalValue = value;
            }
        }

        public static implicit operator HubFloat(float value)
        {
            current.value = value;
            return current;
        }

        public static implicit operator float(HubFloat hub)
        {
            return hub.value;
        }
    }

**Comment by HaloFour** [comment]
@Servail 

Sounds like you found a perfectly good hack to implement the behavior.  I just hope you never need to create two of those types.  Good luck with that!

**Comment by Servail** [comment]
@HaloFour
>create two of those types

Still works. In wrong case I'll just replace "current" with static list and use last element.

**Comment by yaakov-h** [comment]
Maybe an asynclocal list of weak references? 🧐

**Comment by Servail** [comment]
@yaakov-h
Whatever... I don't create them from code anyway, but if, it's one-by-one.

**Reply by iMemento** [reply]
@Servail Would this work?
